File: src/VersionUp/PackageJsonVersionHandler.cs
namespace VersionUp;

using System;
using System.IO;
using System.Text.RegularExpressions;

/// <summary>
/// Handles parsing and updating the version property in package.json files.
/// </summary>
public class PackageJsonVersionHandler : IVersionFileHandler
{
    /// <summary>Matches the <c>"version"</c> property value in a JSON object.</summary>
    private static readonly Regex VersionRegex = new Regex(
        @"""version""\s*:\s*""([^""]+)""",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    /// <inheritdoc />
    public bool CanHandle(string filePath)
    {
        if (string.IsNullOrEmpty(filePath))
        {
            return false;
        }

        string fileName = Path.GetFileName(filePath).ToLowerInvariant();

        return fileName == "package.json";
    }

    /// <inheritdoc />
    public string? GetVersion(string fileContent)
    {
        if (string.IsNullOrWhiteSpace(fileContent))
        {
            return null;
        }

        Match match = VersionRegex.Match(fileContent);

        if (match.Success)
        {
            string value = match.Groups[1].Value;

            return value;
        }

        return null;
    }

    /// <inheritdoc />
    public string UpdateVersion(string fileContent, string newVersion)
    {
        if (string.IsNullOrWhiteSpace(fileContent))
        {
            return $"{{\n  \"version\": \"{newVersion}\"\n}}";
        }

        string result = VersionRegex.Replace(fileContent, match =>
        {
            string rawMatch = match.Value;
            string rawValue = match.Groups[1].Value;

            int index = rawMatch.IndexOf(rawValue, StringComparison.Ordinal);

            if (index >= 0)
            {
                string replaced = rawMatch.Substring(0, index) + newVersion + rawMatch.Substring(index + rawValue.Length);

                return replaced;
            }

            return rawMatch;
        });

        return result;
    }
}

File: vscode-extension/src/handlers/PackageJsonVersionHandler.ts
import * as path from 'path';
import { IVersionFileHandler } from './IVersionFileHandler';

/**
 * Handles parsing and updating the version property in package.json files.
 */
export class PackageJsonVersionHandler implements IVersionFileHandler {
    private static readonly versionRegex = /"version"\s*:\s*"([^"]+)"/gi;

    public canHandle(filePath: string): boolean {
        if (!filePath) {
            return false;
        }

        const fileName = path.basename(filePath).toLowerCase();
        return fileName === 'package.json';
    }

    public getVersion(fileContent: string): string | null {
        if (!fileContent || !fileContent.trim()) {
            return null;
        }

        PackageJsonVersionHandler.versionRegex.lastIndex = 0;
        const match = PackageJsonVersionHandler.versionRegex.exec(fileContent);
        if (match) {
            return match[1];
        }

        return null;
    }

    public updateVersion(fileContent: string, newVersion: string): string {
        if (!fileContent || !fileContent.trim()) {
            return `{\n  "version": "${newVersion}"\n}`;
        }

        PackageJsonVersionHandler.versionRegex.lastIndex = 0;
        return fileContent.replace(PackageJsonVersionHandler.versionRegex, (match, rawValue) => {
            const index = match.indexOf(rawValue);
            if (index >= 0) {
                return match.slice(0, index) + newVersion + match.slice(index + rawValue.length);
            }
            return match;
        });
    }
}
